File: src/Tests/StateUITests/ModalStackTests.swift
// What is presented over the window, as Swift describes it.
//
// A modal stack is an array the author holds, exactly as a navigation path is,
// and it rides as a list of pages under one wrapper node beside the window's
// own page. Coming back there is one report, and it says how many are STILL
// presented - the sheet the reader dragged down has already gone.
//
// What the renderer does with it is next door, in the C# ModalStackTests.

import StateUIWireProbe
import XCTest
@testable import StateUI

/// What an application presents over itself. An enum, because the destination
/// is a `switch` and the compiler is what says every case has a page.
private enum Sheet: Hashable {
    case settings
    case about
}

/// The page underneath, which is what presents.
private struct HomePage: ContentPage {
    @Binding var sheets: [Sheet]

    var title: String? { "Home" }

    var content: Element {
        Button("Settings").onClicked { sheets.append(.settings) }
    }
}

/// A presented page. It carries its own way out, because a modal covers the
/// bars as well as the content and there is nothing else to close it with.
private struct SheetPage: ContentPage {
    @Binding var sheets: [Sheet]

    let name: String

    var title: String? { name }
    var modalPresentationStyle: UIModalPresentationStyle? { .pageSheet }

    var content: Element {
        Button("Close").onClicked { sheets.removeLast() }
    }
}

/// The window under test, over whatever state is lent to it.
private struct TestWindow: Window {
    let sheets: Binding<[Sheet]>

    /// The stack the window shows, for the one test that puts a navigation page
    /// under the sheets. Nil is the plain home page.
    var path: Binding<[Int]>?

    var title: String?

    var modalStack: ModalStack? {
        ModalStack(sheets) { sheet in
            switch sheet {
            case .settings: SheetPage(sheets: sheets, name: "Settings")
            case .about: SheetPage(sheets: sheets, name: "About")
            }
        }
    }

    var content: Page {
        guard let path else { return HomePage(sheets: sheets) }

        return NavigationPage(path) {
            HomePage(sheets: sheets)
        } destination: { _ in
            HomePage(sheets: sheets)
        }
        .title("Diary")
    }
}

/// The window under test, over whatever state is lent to it.
private func window(_ sheets: Binding<[Sheet]>) -> Window {
    TestWindow(sheets: sheets)
}

final class ModalStackTests: XCTestCase {
    // MARK: - What goes out

    /// A window with nothing presented still says so: the list is there and it
    /// is empty, which is what tells the host to dismiss whatever it is still
    /// holding.
    func testAWindowWithNothingPresentedCarriesAnEmptyList() {
        let sheets = State<[Sheet]>([])

        let node = window(sheets.projectedValue).body.built

        XCTAssertEqual(node.children.map { $0.type.name }, ["ContentPage", "ModalStack"])
        XCTAssertEqual(node.children.last?.children.count, 0)
    }

    /// One presented page is one child of the list, and its identity carries
    /// its DEPTH as well as its value - the rule a navigation stack follows,
    /// for the reason a stack has it: two identical sheets are two pages.
    func testAPresentedPageIsAChildOfTheListWearingItsDepth() {
        let sheets = State<[Sheet]>([.settings, .about])

        let list = window(sheets.projectedValue).body.built.children.last

        XCTAssertEqual(list?.children.map { $0.id }, ["0/settings", "1/about"])
        XCTAssertEqual(list?.children.map { $0.built.props["title"] },
                       [.string("Settings"), .string("About")])
    }

    /// Presenting is appending, and that is the whole of it: the tap on the
    /// page underneath writes the array, and the next render carries a page
    /// that was not there before.
    func testPresentingIsAppendingToTheArray() {
        let sheets = State<[Sheet]>([])
        let renders = Renders()

        let first = renders.render(window(sheets.projectedValue).body)
        let button = first.children.first?.children.first

        XCTAssertTrue(renders.fire(button?.events?["clicked"] ?? -1))
        XCTAssertEqual(sheets.wrappedValue, [.settings])

        let patch = renders.render(window(sheets.projectedValue).body)
        let list = patch.children.first { $0.type == "ModalStack" }

        XCTAssertEqual(list?.children.map { $0.id }, [.manual("0/settings")])
    }

    /// And closing is the array getting shorter. The button is on the SHEET,
    /// which is where it has to be: a modal covers the bar the page underneath
    /// would have offered.
    func testClosingIsTheArrayGettingShorter() {
        let sheets = State<[Sheet]>([.settings])
        let renders = Renders()

        let first = renders.render(window(sheets.projectedValue).body)
        let close = first.children.first { $0.type == "ModalStack" }?
            .children.first?.children.first

        XCTAssertTrue(renders.fire(close?.events?["clicked"] ?? -1))
        XCTAssertEqual(sheets.wrappedValue, [])

        let patch = renders.render(window(sheets.projectedValue).body)

        XCTAssertEqual(patch.children.first { $0.type == "ModalStack" }?.children.count, 0)
    }

    /// The style a page is drawn with is the PAGE's own property, not the
    /// stack's - so a sheet knows what it looks like wherever it is presented
    /// from, and a page pushed onto a navigation stack simply carries a
    /// property nothing reads.
    func testHowAPageIsPresentedIsThePagesOwnProperty() {
        let sheets = State<[Sheet]>([.settings])

        let sheet = window(sheets.projectedValue).body.built.children.last?.children.first

        XCTAssertEqual(sheet?.built.props["modalPresentationStyle"], .enumeration(3), "pageSheet")
    }

    /// A page the library CONSTRUCTS says the same thing by modifier, which is
    /// the shape of a sheet on iOS: a whole navigation stack presented as a
    /// card, with a bar and a way out of its own.
    func testAConstructedPageSaysItByModifier() {
        let path = State<[Int]>([])

        let node = NavigationPage(path.projectedValue) {
            SheetPage(sheets: State<[Sheet]>([]).projectedValue, name: "Settings")
        } destination: { _ in
            SheetPage(sheets: State<[Sheet]>([]).projectedValue, name: "Deeper")
        }
        .modalPresentationStyle(.formSheet)
        .body
        .built

        XCTAssertEqual(node.props["modalPresentationStyle"], .enumeration(1), "formSheet")
    }

    // MARK: - What comes back

    /// The reader's own way out - an iOS sheet dragged down, Android's system
    /// back. The payload is what SURVIVED, and the array is truncated to it.
    func testADismissalTruncatesTheArray() {
        let sheets = State<[Sheet]>([.settings])
        let renders = Renders()

        let patch = renders.render(window(sheets.projectedValue).body)

        XCTAssertTrue(renders.fire(patch.events?["modalPopped"] ?? -1, with: [.number(0)]))
        XCTAssertEqual(sheets.wrappedValue, [])
    }

    /// Only the top one went: a report of one surviving over a stack of two
    /// leaves the first sheet presented.
    func testADismissalOfTheTopLeavesWhatIsUnderIt() {
        let sheets = State<[Sheet]>([.settings, .about])
        let renders = Renders()

        let patch = renders.render(window(sheets.projectedValue).body)

        XCTAssertTrue(renders.fire(patch.events?["modalPopped"] ?? -1, with: [.number(1)]))
        XCTAssertEqual(sheets.wrappedValue, [.settings])
    }

    /// A report that would LENGTHEN the array is refused - it has been
    /// overtaken by something this side already did, and obeying it would put a
    /// dismissed sheet back on the screen.
    func testAReportThatWouldPresentSomethingAgainIsRefused() {
        let sheets = State<[Sheet]>([])
        let renders = Renders()

        let patch = renders.render(window(sheets.projectedValue).body)

        XCTAssertTrue(renders.fire(patch.events?["modalPopped"] ?? -1, with: [.number(2)]))
        XCTAssertEqual(sheets.wrappedValue, [])
    }

    /// A payload of the wrong shape leaves the array alone.
    func testAValueOfTheWrongKindLeavesTheArrayAlone() {
        let sheets = State<[Sheet]>([.settings])
        let renders = Renders()

        let patch = renders.render(window(sheets.projectedValue).body)

        XCTAssertTrue(renders.fire(patch.events?["modalPopped"] ?? -1, with: [.string("0")]))
        XCTAssertEqual(sheets.wrappedValue, [.settings])
    }

    // MARK: - The contract the C# side reads

    /// The whole thing, written down: a window whose page is a navigation
    /// stack, with two pages presented over all of it.
    func testTheModalStackIsWrittenDown() throws {
        let sheets = State<[Sheet]>([.settings, .about])
        let path = State<[Int]>([1])
        let differ = Differ()

        let tree = TestWindow(
            sheets: sheets.projectedValue,
            path: path.projectedValue,
            title: "StateUI").body

        let result = differ.reconcile(nil, with: tree)
        let bytes = Wire.encode(result.patch, generation: 1, dictionary: WireDictionary())

        try Fixtures.check(
            bytes,
            sidecar: WireProbe.dumpMessage(bytes, names: WireNames()),
            against: "pages/ModalStack")
    }
}

File: src/Tests/StateUIRuntime.Tests/ModalStackTests.cs
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

// What is presented over the window, as the renderer does it.
//
// Swift describes the whole modal stack as an ARRANGED children list under one
// wrapper node, and this side brings MAUI's own modal stack to it - pushing
// what is described and dismissing what is not, top first, because the top is
// the only end of a modal stack any platform lets go of. The one thing that
// travels back is a dismissal the reader made.
//
// What Swift puts on the wire is next door, in the Swift ModalStackTests.
using StateUI.Runtime.Protocol;
using StateUI.Runtime.Rendering;

// The one name out of the iOS platform-specific namespace this file reads.
using iOSPage = Microsoft.Maui.Controls.PlatformConfiguration.iOSSpecific.Page;
using UIModalPresentationStyle =
    Microsoft.Maui.Controls.PlatformConfiguration.iOSSpecific.UIModalPresentationStyle;

namespace StateUI.Runtime.Tests;

public class ModalStackTests
{
    /// <summary>The handler id the window's node says a dismissal reports to.</summary>
    private const int Popped = 9;

    /// <summary>
    /// A window with a page, a page renderer over it, and the subscription
    /// <see cref="StateUIWindow"/> makes - which is where MAUI raises a
    /// modal's departure, the modal stack being the window's own.
    /// </summary>
    /// <remarks>
    /// A plain MAUI window rather than a StateUIWindow, so that the reports
    /// land in a test double: a StateUIWindow builds a session of its own,
    /// whose dispatch goes to the native side. The fixture at the bottom is
    /// the one test that wants the real thing.
    /// </remarks>
    private static (SwiftPages Pages, Host Host, Window Window) Renderer()
    {
        var host = new Host();
        var pages = new SwiftPages(
            host.Renderer,
            (message, exception) => Assert.Fail($"{message}\n{exception}"));

        var window = new Window { Page = new ContentPage { Title = "Home" } };

        // What Apply does for the window's own node: the handler ids a
        // dismissal is reported with are the window's, so the window has to be
        // tracked before anything can be said about it.
        host.Renderer.Track(window, Host.Parse(
            $"{{\"id\":1,\"type\":\"Window\",\"events\":{{\"modalPopped\":{Popped}}}}}"));

        window.ModalPopped += (_, _) => pages.ModalWasPopped();

        return (pages, host, window);
    }

    /// <summary>One presented page, as Swift writes it.</summary>
    /// <remarks>
    /// The style is a member of a closed vocabulary, so it goes on as a number
    /// - <see cref="Host.Member"/> spells one - and the parameter is the mirror
    /// itself, which is what stops a sheet asking for a style no member names.
    /// </remarks>
    private static string Sheet(
        string identity, string title, SwiftUIModalPresentationStyle? style = null) =>
        $"{{\"id\":\"{identity}\",\"type\":\"ContentPage\",\"props\":{{\"title\":\"{title}\""
        + (style is null ? "" : $",\"modalPresentationStyle\":{Host.Member(style.Value)}")
        + "},\"arranged\":true,\"children\":"
        + $"[{{\"id\":1,\"type\":\"Label\",\"props\":{{\"text\":\"{title}\"}}}}]}}";

    /// <summary>The modal stack node over the sheets named, innermost first.</summary>
    private static SwiftNode Stack(params string[] sheets) => Host.Parse(
        $"{{\"id\":2,\"type\":\"ModalStack\",\"arranged\":true,\"children\":[{string.Join(",", sheets)}]}}");

    /// <summary>The titles of what is presented, innermost first.</summary>
    private static string[] Presented(Window window) =>
        [.. window.Navigation.ModalStack.Select(page => page.Title ?? "")];

    // ---- What an arrangement does ------------------------------------------

    /// <summary>A window with nothing presented over it is the ordinary one.</summary>
    [Fact]
    public void AnEmptyListPresentsNothing()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack());

        Assert.Empty(window.Navigation.ModalStack);
    }

    [Fact]
    public void AListOfOnePresentsIt()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/settings", "Settings")));

        Assert.Equal(["Settings"], Presented(window));
    }

    /// <summary>
    /// Modals stack, which is why this is a list and not a page: a sheet may
    /// present a sheet, and the platforms all allow it.
    /// </summary>
    [Fact]
    public void ALongerListPresentsOverWhatIsAlreadyThere()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A")));
        Page first = window.Navigation.ModalStack[0];

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A"), Sheet("1/b", "B")));

        Assert.Equal(["A", "B"], Presented(window));
        Assert.Same(first, window.Navigation.ModalStack[0]);
    }

    /// <summary>A shorter list dismisses, from the top.</summary>
    [Fact]
    public void AShorterListDismissesTheTop()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A"), Sheet("1/b", "B")));
        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A")));

        Assert.Equal(["A"], Presented(window));
    }

    /// <summary>And an empty one dismisses everything there is.</summary>
    [Fact]
    public void AnEmptyListDismissesAllOfThem()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A"), Sheet("1/b", "B")));
        pages.ApplyModals(window.Navigation, window, Stack());

        Assert.Empty(window.Navigation.ModalStack);
    }

    /// <summary>
    /// A sheet SWAPPED for another at the same depth: the one that was there
    /// goes and the new one is presented, which the loop reaches by dismissing
    /// down to the common part and pushing back up.
    /// </summary>
    [Fact]
    public void ADifferentSheetAtTheSameDepthReplacesIt()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A")));
        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/b", "B")));

        Assert.Equal(["B"], Presented(window));
    }

    /// <summary>
    /// A patch about a presented page keeps the PAGE - its controls, its scroll
    /// offset, whatever the reader typed into it - the way every other
    /// arrangement keeps what it was not told to change.
    /// </summary>
    [Fact]
    public void APatchAboutASheetKeepsThePage()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/settings", "Settings")));
        Page sheet = window.Navigation.ModalStack[0];

        pages.ApplyModals(window.Navigation, window, Host.Parse(
            "{\"id\":2,\"type\":\"ModalStack\",\"children\":["
            + "{\"id\":\"0/settings\",\"type\":\"ContentPage\",\"props\":{\"title\":\"Options\"}}]}"));

        Assert.Same(sheet, window.Navigation.ModalStack[0]);
        Assert.Equal(["Options"], Presented(window));
    }

    /// <summary>
    /// A page REBUILT under an unchanged identity - which is what
    /// <c>replace</c> means - is presented in place of the one that was there,
    /// even though no arrangement arrived to say so.
    /// </summary>
    [Fact]
    public void ARebuiltSheetIsPresentedInPlaceOfTheOldOne()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/settings", "Settings")));
        Page sheet = window.Navigation.ModalStack[0];

        pages.ApplyModals(window.Navigation, window, Host.Parse(
            "{\"id\":2,\"type\":\"ModalStack\",\"children\":["
            + "{\"id\":\"0/settings\",\"type\":\"ContentPage\",\"replace\":true,"
            + "\"props\":{\"title\":\"Settings\"}}]}"));

        Assert.NotSame(sheet, window.Navigation.ModalStack[0]);
        Assert.Single(window.Navigation.ModalStack);
    }

    /// <summary>A child that is not a page is reported, not skipped.</summary>
    [Fact]
    public void AChildThatIsNotAPageIsReported()
    {
        var host = new Host();
        List<string> failures = [];
        var pages = new SwiftPages(host.Renderer, (message, _) => failures.Add(message));
        var window = new Window { Page = new ContentPage() };

        pages.ApplyModals(window.Navigation, window, Host.Parse(
            "{\"id\":2,\"type\":\"ModalStack\",\"arranged\":true,\"children\":["
            + "{\"id\":3,\"type\":\"Label\",\"props\":{\"text\":\"not a page\"}}]}"));

        Assert.Contains(failures, message => message.Contains("modal stack"));
    }

    // ---- How it is drawn ----------------------------------------------------

    /// <summary>
    /// The presentation style is the PAGE's own property, so it reaches the
    /// page through the same chrome every page goes through.
    /// </summary>
    /// <remarks>
    /// UIKit is the only platform that reads it. Setting it here on a test
    /// machine is what the renderer does on Android too: an ordinary bindable
    /// property of MAUI's shared assembly, which nothing looks at there.
    /// </remarks>
    [Fact]
    public void TheStyleReachesThePresentedPage()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window,
            Stack(Sheet("0/settings", "Settings", SwiftUIModalPresentationStyle.PageSheet)));

        Assert.Equal(
            UIModalPresentationStyle.PageSheet,
            iOSPage.GetModalPresentationStyle(window.Navigation.ModalStack[0]));
    }

    /// <summary>
    /// Every style the Swift side can write is a style this side knows - a
    /// number no member names leaves MAUI's own default standing.
    /// </summary>
    /// <remarks>
    /// The member arrives as an <c>int</c> because the mirror is internal and a
    /// public test method cannot take one; the cast back at the top of the body
    /// is where it becomes a member again.
    /// </remarks>
    /// <param name="member">The style's number, as <c>SwiftUIModalPresentationStyle</c>.</param>
    /// <param name="expected">The UIKit style it has to reach.</param>
    [Theory]
    [InlineData((int)SwiftUIModalPresentationStyle.FullScreen, UIModalPresentationStyle.FullScreen)]
    [InlineData((int)SwiftUIModalPresentationStyle.FormSheet, UIModalPresentationStyle.FormSheet)]
    [InlineData((int)SwiftUIModalPresentationStyle.Automatic, UIModalPresentationStyle.Automatic)]
    [InlineData((int)SwiftUIModalPresentationStyle.PageSheet, UIModalPresentationStyle.PageSheet)]
    [InlineData(
        (int)SwiftUIModalPresentationStyle.OverFullScreen, UIModalPresentationStyle.OverFullScreen)]
    [InlineData((int)SwiftUIModalPresentationStyle.Popover, UIModalPresentationStyle.Popover)]
    public void EveryStyleIsUnderstood(int member, UIModalPresentationStyle expected)
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window,
            Stack(Sheet("0/settings", "Settings", (SwiftUIModalPresentationStyle)member)));

        Assert.Equal(expected, iOSPage.GetModalPresentationStyle(window.Navigation.ModalStack[0]));
    }

    // ---- What comes back ----------------------------------------------------

    /// <summary>
    /// The reader's own way out - an iOS sheet dragged down, Android's system
    /// back - is reported as what SURVIVED, so the bound array can be truncated
    /// to it.
    /// </summary>
    [Fact]
    public async Task AReaderDismissingIsReported()
    {
        (SwiftPages pages, Host host, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/settings", "Settings")));

        host.Dispatched.Clear();

        // What a drag down does, in the one way a headless test can do it.
        await window.Navigation.PopModalAsync();

        Assert.Equal([(Popped, "0")], host.Dispatched);
    }

    /// <summary>
    /// And what SWIFT asked for is not reported back - not during the message,
    /// and not a turn later either, which is when every report is made.
    /// </summary>
    [Fact]
    public void WhatSwiftAskedForIsNotReportedBack()
    {
        (SwiftPages pages, Host host, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/settings", "Settings")));

        host.Dispatched.Clear();
        TestDispatcher.Hold();

        using (StateUIRenderer.Suppressed applying = host.Renderer.Applying())
        {
            pages.ApplyModals(window.Navigation, window, Stack());
        }

        TestDispatcher.Drain();

        Assert.Empty(window.Navigation.ModalStack);
        Assert.Empty(host.Dispatched);
    }

    /// <summary>
    /// A dismissal that lands where Swift already believes says nothing - the
    /// same guard the navigation stack's depth has, and what keeps a sheet
    /// closed by its own button to exactly one render.
    /// </summary>
    [Fact]
    public async Task ADismissalOfWhatSwiftAlreadyClosedSaysNothing()
    {
        (SwiftPages pages, Host host, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/settings", "Settings")));
        pages.ApplyModals(window.Navigation, window, Stack());

        host.Dispatched.Clear();

        // Nothing left to dismiss, and nothing to say about it.
        await Task.Yield();

        Assert.Empty(host.Dispatched);
    }

    /// <summary>
    /// Only the top of a stack of two: what is reported is the DEPTH that
    /// survived, so an application that presented two sheets learns it still
    /// has one.
    /// </summary>
    [Fact]
    public async Task ADismissalOverAStackReportsWhatSurvived()
    {
        (SwiftPages pages, Host host, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A"), Sheet("1/b", "B")));

        host.Dispatched.Clear();

        await window.Navigation.PopModalAsync();

        Assert.Equal([(Popped, "1")], host.Dispatched);
    }

    // ---- What the renderer forgets ------------------------------------------

    [Fact]
    public void ForgettingBuildsThemAgain()
    {
        (SwiftPages pages, _, Window window) = Renderer();

        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A")));
        Page sheet = window.Navigation.ModalStack[0];

        pages.Forget();
        pages.ApplyModals(window.Navigation, window, Stack(Sheet("0/a", "A")));

        Assert.NotSame(sheet, window.Navigation.ModalStack[^1]);
    }

    // ---- The fixture, which is the contract --------------------------------

    /// <summary>
    /// The bytes the Swift tests wrote, through a real window: a navigation
    /// stack pushed one deep, with two sheets presented over all of it.
    /// </summary>
    /// <remarks>
    /// The whole path, which is what this one is for - the window recognizing
    /// the modal stack among its children, and reaching it AFTER the page it
    /// also carries.
    /// </remarks>
    [Fact]
    public void TheFixturePresentsWhatSwiftDescribed()
    {
        var window = Host.Window();

        window.Apply(
            Host.Parse(Fixtures.ReadBytes("pages/ModalStack.bin")), true);

        Assert.IsType<NavigationPage>(window.Page);
        Assert.Equal(["Settings", "About"], Presented(window));

        Assert.Equal(
            UIModalPresentationStyle.PageSheet,
            iOSPage.GetModalPresentationStyle(window.Navigation.ModalStack[0]));
    }
}
